src/addrman.h

```diff
@@ -262,7 +297,7 @@ friend class CAddrManTest;
     bool Add_(const CAddress &addr, const CNetAddr& source, int64_t nTimePenalty) EXCLUSIVE_LOCKS_REQUIRED(cs);
 
     //! Mark an entry as attempted to connect.
-    void Attempt_(const CService &addr, bool fCountFailure, int64_t nTime) EXCLUSIVE_LOCKS_REQUIRED(cs);
+    void AttemptWithLockHeld(const CService &addr, bool fCountFailure, int64_t nTime) EXCLUSIVE_LOCKS_REQUIRED(cs);
```

Comment:
I don't think we need to expose this as a public method just for the addrman tests. Just release the lock in that test before calling `Attempt()`:

```diff
--- a/src/test/addrman_tests.cpp
+++ b/src/test/addrman_tests.cpp
@@ -73,13 +73,12 @@ public:
     // Simulates connection failure so that we can test eviction of offline nodes
     void SimConnFail(const CService& addr)
     {
-         LOCK(cs);
-         int64_t nLastSuccess = 1;
-         Good_(addr, true, nLastSuccess); // Set last good connection in the deep past.
+         // Set last good connection in the deep past.
+         WITH_LOCK(cs, {Good_(addr, true, /* nLastSuccess= */ 1);});
 
          bool count_failure = false;
          int64_t nLastTry = GetAdjustedTime()-61;
-         AttemptWithLockHeld(addr, count_failure, nLastTry);
+         Attempt(addr, count_failure, nLastTry);
```

The unit test is single threaded, so there's no need to hold the mutex between `Good_()` and `Attempt()`.

Reply:
Thanks! Done in bc3694a49a93b9ae6f8cacfd0d5b1def5f09b004.